### DISCUSSION TITLE
[IDEA] Retry mechanism and concurrent download through go channels

### DISCUSSION THREAD

**Discussion by arvpo** [arvpo/blazer#50] [Ideas]
I have attached sample code and a small diagram describing the architecture briefly

sample-code: https://github.com/arvyshka/blazer/blob/experiments/experiments/worker.go

![image](https://user-images.githubusercontent.com/4346153/128914620-f519dd23-225a-40d0-bedf-f963ae5656ab.png)

sQ - successqueue
jQ - jobqueue
lQ - last queue (vertict queue)
mR - max retry
∞ - iinfite for loop 
but the retry logic i didn't implement as shown in the diagram (in the sample code), just retry queue with some mock logic

**Comment by arvpo** [comment]
if you run the sample code, it simulates a number of jobs running concurrently with some jobs failing and being pushed to retry queue, then those jobs also run and complete the download process. if the max retry is exceeded, we need to tell the verdict queue, that download incomplete and hence we will return bool accordingly.. @setraj If it is unclear, i will explain in detail. maybe in a zoom call

**Reply by setraj** [reply]
To start simple, why can't we let each goroutine take care of its failed segment's retry process?
Basically, when a segment download fails, the corresponding goroutine retries for max number of times.

**Reply by arvpo** [reply]
> To start simple, why can't we let each goroutine take care of its failed segment's retry process?
> Basically, when a segment download fails, the corresponding goroutine retries for max number of times.

yes, right now it works like that. Did you see the code, there, it retries after some time
